### PR TITLE
validate checkpoint

### PR DESCRIPTION
**Pull request by zhiqiangxu**
validate validators set for checkpoint, otherwise evil nodes can do bad things.

### CODE REVIEW

File: consensus/bor/bor.go [hunk 1]
```diff
@@ -399,6 +399,22 @@ func (c *Bor) verifyCascadingFields(chain consensus.ChainHeaderReader, header *t
 		return err
 	}
 
+	if (number+1)%c.config.Sprint == 0 {
```
Comment (jdkanani) [inline consensus/bor/bor.go:402]: It's already checking it from previous block's header and that header is built from the `BorValidatorSet` contract. https://github.com/maticnetwork/bor/blob/master/consensus/bor/bor.go#L403-L417

File: consensus/bor/bor.go [hunk 2]
```diff
@@ -399,6 +399,22 @@ func (c *Bor) verifyCascadingFields(chain consensus.ChainHeaderReader, header *t
...
+		newValidators, err := c.GetCurrentValidators(header.ParentHash, number+1)
```
Comment (jdkanani) [inline consensus/bor/bor.go:403]: I am not sure if this works in full sync mode. That was the reason we tweaked it to check validators from previous header.

**Review by jdkanani — COMMENTED** [review]
Thanks a lot for the PR. Added few comments.

File: consensus/bor/bor.go (hunk 1 above)
Comment (zhiqiangxu) [reply consensus/bor/bor.go:402]: the first block of a sprint will takes validators from last checkpoint, without validation.

File: consensus/bor/bor.go (hunk 2 above)
Comment (zhiqiangxu) [reply consensus/bor/bor.go:403]: I think you mean the check of `validateHeaderExtraField`, but it only checks length, doesn't validate against the actual validators.

**Comment by temaniarpit27** [comment]
@zhiqiangxu Please provide the latest update for this

**Comment by temaniarpit27** [comment]
Closing this for now

**Pull request closed without merging**